Refactor hydro max power TS consistency

guilpier-code opened the pull request:
What was done (mainly) :
- [x] Factorizing the code that make time series collections of same size
- [x] Changing TS counts for : 
    - **ror**, **storage**, **mingen** ,
    - max power **gen** and **pump**
  
  from a **public** data member into a **private** data member. Indeed, one should not be able to change this TS count without resizing the associated TS accordingly.

Some other things done : 
- [x] Remove the exception raising, not necessary : the **gotFatalError** (set to true) will trigger an exception anyway.
- [x] class **NbTsComparer** : is renamed (into **pairOfTSsizes**), generalized to any numbers (not just TS numbers) and moved elsewhere 
- [x] Class **TsActions** removed (useless)

guilpier-code committed:
Refactor max power TS checks : several things (see commit message)

This commit is large, it should be shorter, but the long compilation time forces to gather the devs in one piece.
- Remove the exception raising, not necessary : the gotFatalError (set to true) will trigger an exception anyway.
- class NbTsComparer : is generalized to any time series pair :
   + to isolate it from business context
   + to have a chance to re-use it in another context
- Move both classes NbTsComparer and TsActions outside class DataSeriesHydro, for same reasons as above
- Rename TsActions::handleBothGreaterThanOne (unclear) into something more meaningful
- Consequences of previous changes in tests

guilpier-code committed:
Refactor max power TS checks : several things (see commit message)

TsActions::areaToInvalidate :
- Remove areaToInvalidate from class TsActions (it has nothing to do in here) and make it a free function
- Extract areaToInvalidate calls from TsActions::resizeWhenOneTS, as calling it here not related to TsActions

guilpier-code committed:
Refactor max power TS checks : simplify TsActions::resizeWhenOneTS

guilpier-code committed:
Refactor max power TS checks : remove a useless argument to free function invalidateArea

guilpier-code committed:
Refactor max power TS checks : changing invalidateArea's signature and content

In this context, pointer area (as argument) is necessarily not null, because it is used way before we come to this point.
So :
- We make area a reference
- We stop testing the area address (we known it's not null)
- We remove the dead code (when area address is null)

guilpier-code committed:
Refactor max power TS checks : several things (see commit message)

- Move class NumberComparison to separate source files
- Remove return type to function postProcessMaxPowerTS : no more boolean returned, but void.
  Returning boolean is for loading tasks, here we check and possibly change already loaded data.
  We adapt tests, as a consequence
- Removing class TsActions, not useful
-

nikolaredstork reviewed (commented):
There are no numberComparison .h and .cpp files. Also, maybe I'm wrong, in **src/libs/antares/study/parts/parts.h** file, you probably have to include numberComparison.h file.

guilpier-code committed:
Refactor max power TS checks : add forgotten files + renaming

guilpier-code committed:
Refactor max power TS checks : remove case where both TS matrices (pump + gen) have null width, impossible

guilpier-code committed:
Refactor max power TS checks : just removing useless comments

guilpier-code committed:
Refactor max power TS checks : a bot of cleaning

nikolaredstork commented on src/libs/antares/study/parts/hydro/pair-of-integers.h:
```diff
@@ -0,0 +1,17 @@
+    unsigned int numberOfTS_2_;
+    unsigned int numberOfTSsup_{ 0 };
+    unsigned int numberOfTSinf_{ 0 };
+};
```
For the ror, hydro-storage and mingen there is three number comparison, so we should go with class that can receive variable list of integers, so that we can use it for both cases:

-  Maximum Generation/Pumping 2 integers comparison.
- ROR, Hydro-Storage and Minimum Generation 3 integers comparison.
![image](https://github.com/AntaresSimulatorTeam/Antares_Simulator/assets/131546833/b2445533-76de-4d5c-bb4b-f844a364ca68)

guilpier-code replied:
Not sure it would be useful, and here is why (in my opinion) :
- Let's take an exemple from another field : sort algorithm. From a comparison between 2 elements, sort algorithm manages to sort the elements of a list.
Here, the idea would be the same : use the comparison between 2 elements to compare many of them.
So we should keep this class as it is, and build on it.
- As you said, we have the **ror**, **storage** and **mingen** time series, and we have to "equalize" their sizes. It's currently done in function **DataSeriesHydro::EqualizeGenerationTSsizes**. As you can see in this function's body, there is some code duplication : we have 2 blocks very much alike. This is usually a sign that we should refactor to remove the code duplication. But I spent some time to get to this results on reading the hydro TS, and there's still things to do, apart from removing this code duplication which is, in fact the subject you're talking about here.
To me, we should first try to merge this as it is, create a gitHub issue if we're not happy, and fix this later.

guilpier-code committed:
Refactor max power TS checks : internalize count of generation time series (ror, storage, mingen)

We should not be able to change a TS count without resizing the TS.
This was allowed when because this count was a public data member.
So we make it a private data member, and change it only when resizing the associated data.

nikolaredstork commented on src/libs/antares/study/area/list.cpp:
```diff
@@ -1641,10 +1641,7 @@ void AreaList::removeLoadTimeseries()
-        area.hydro.series->storage.reset(1, DAYS_PER_YEAR);
-        area.hydro.series->mingen.reset(1, HOURS_PER_YEAR);
-        area.hydro.series->count = 1;
+        area.hydro.series->reset();
```
**void AreaList::removeHydroTimeseries()** function is used to reset TS's when TS generator is enabled. 
![image](https://github.com/AntaresSimulatorTeam/Antares_Simulator/assets/131546833/81853590-7be7-4a38-9b14-d6f47dbe9ea7)

Maximum Generation/Pumping can't be generated via TS generator, so do we have to reset them when TS generator is enabled for hydro?

guilpier-code replied:
As you just said, **AreaList::removeHydroTimeseries()** is only by the historic **GUI**. 
This GUI will disappear (the sooner the better), and will be fully replaced with the web version.
So this function will be removed as well.
But you're right, some extra things are done that we don't need. 
**Unless we introduce a bug in the GUI**, the work required to have it clean is not worth it.

guilpier-code committed:
Refactor max power TS checks : renaming

guilpier-code committed:
Refactor max power TS checks : renaming

guilpier-code committed:
Refactor max power TS checks : adapt CI Window workflow to new runner (about python)

guilpier-code committed:
Refactor max power TS checks : define and use of function EqualizeTSsize(...)

guilpier-code committed:
Refactor max power TS checks : rearrange DataSeriesHydro::loadFromFolder, but no major change

guilpier-code committed:
Refactor max power TS checks : remove the case where TS (ror, storage, mingen) have a zero size. Impossible case.

guilpier-code committed:
Refactor max power TS checks : fix bugs on function EqualizeTSsize

guilpier-code committed:
Refactor max power TS checks :  use maxPowerTScount() whenever possible

guilpier-code committed:
Refactor max power TS checks : removing the setter for max power TS number

guilpier-code committed:
Refactor max power TS checks : fix a bug due to removing the setter for max power TS number

guilpier-code committed:
Refactor max power TS checks : gather loading and equalizing hydro mingen TS

guilpier-code committed:
Refactor max power TS checks : split load of ROR, STORAGE and MINGEN

guilpier-code committed:
Refactor max power TS checks : loading Max Power TS : using a generic loading function

guilpier-code committed:
Refactor max power TS checks : fix a bug on loading inflows (height ont correct)

guilpier-code committed:
Refactor max power TS checks : correction after (small) merge

guilpier-code committed:
Refactor max power TS checks : fixing bug on loading inflows was wrong. Fixing it for real now.

guilpier-code committed:
Refactor max power TS checks : remove useless checkMinGenTsNumber

guilpier-code committed:
Refactor max power TS checks : cleaning on the very ugly setHydroModulability (too many args, body too big)

nikolaredstork commented on src/libs/antares/study/parts/hydro/hydromaxtimeseriesreader.cpp:
```diff
@@ -122,59 +122,15 @@ void HydroMaxTimeSeriesReader::copyDailyMaxPumpingEnergy(Area& area) const
     ret = loadDailyMaxPowersAndEnergies(folder, area, usedBySolver) && ret;
     copyDailyMaxEnergy(area);
-    copyDailyMaxPowerAsHourly(area);
+    area.hydro.series->buildMaxPowerFromDailyTS(dailyMaxPumpAndGen[genMaxP], dailyMaxPumpAndGen[pumpMaxP]);
```
The whole idea was to wrap around necessary code which will provide support for old studies, and also if in future we don't need that support anymore, we could just erase `HydroMaxTimeSeriesReader` class and it's usage in **list.cpp** file. With current situation `DataSeriesHydro` class have functions which are going to be used just if heder version variable is less than 870 and with upgrading a study these functions are useless  for `DataSeriesHydro` class, so I don't think that functions:

- `buildMaxPowerFromDailyTS`
- `ConvertDailyTSintoHourlyTS`

should be part of the DataSeriesHydro class.

guilpier-code replied:
Here was my idea on that matter : 
- I saw what you were trying to do : **isolate** and **externalize** the reading of the daily TS to turn them into some hourly TS. The idea of isolating looks really good to me. But reading these daily TS is still reading hydro TS, and so it should be done inside a reading function or class, for logical and readability reasons.
- **Externalize** : what you did was **externalizing** the daily TS reading, that is : reading these TS outside the TS reader itself (**HydroMaxTimeSeriesReader** is totally external  to **DataSeriesHydro**). This is what is called a **feature envy**. A consequence of this is that you're forced to supply your reader (**DataSeriesHydro**) with a **public** setter **setNbTimeSeriesSup(...)**, and that the real TS  sizes (inside the reader) are not synchronized automatically with the **number of TS** (**nbTimeSeriesSup_**) : you can set this **number of TS** without modifying the size of your TSs. And this is a problem to me.
- **isolate** : as I said, isolating **IS** very much suitable. But it has to be done inside the reader : the reader delegates the reading of daily TS to an object of which reading daily TS is its specialty.

The design of the reader is still not currently ideal (we hope it will be improved soon).
Following what I just said, here is a rough idea of what it could be (at least it would be much better than today) : 
```C++
// list.cpp
...
HydroTSfetcher hydroTSfetcher(studyVersion, 
                             pathToInputDir, 
                             area, 
                             parameters.timeSeriesToRefresh, 
                             /* other args ? */);
ret = hydroTSfetcher.go() && ret;
``` 
As you can, there 's only one object / interface to load all TS collections (for a given area). 
Inside the fetcher, we would build and run sub-objects that would **read** or **generate** time series (depending on the value of `parameters.timeSeriesToRefresh`), and that would do it differently depending of the study version. All internal classes would be a lot smaller than the current **DataSeriesHydro**.

But all this is for another time.

nikolaredstork replied:
Thanks for the explanation.

nikolaredstork commented on src/libs/antares/study/parts/hydro/series.h:
```diff
@@ -144,108 +151,34 @@ class DataSeriesHydro
-    public:
-        NbTsComparer(uint32_t nbOfGenPowerTs, uint32_t nbOfPumpPowerTs);
-
-        bool bothZeros() const;
```
Function `bothZeros `is not useless as it seems. When we are loading study via GUI, matrices ror, storage, mingen will not be loaded and their width are 0. In current version we have excluded bothZeros function, so every time on loading study via GUI, code will enter `EqualizeTSsize` function, and because both width are 0, if condition `pairOfTSsizes.same() = true` and function `EqualizeTSsize` will return 0.

![image](https://github.com/AntaresSimulatorTeam/Antares_Simulator/assets/131546833/489fc79f-b9c1-48dc-a851-f38402df5167)


. Maybe we don't need bothZero function, we just have to use usedBySolver bool variable in `EqualizeGenerationTSsizes` function like this:

![image](https://github.com/AntaresSimulatorTeam/Antares_Simulator/assets/131546833/5a90b104-de6a-4e77-a917-3cce5ca9ab4f)

Current version will not break simulation, whether it is loaded via GUI or solver, but it is not necessary to check matrices if we know that their width are 0 when we loading study via GUI.

**NOTE:** If you are wondering what will happen to the MaxPower TS (maxHourlyGenPower and maxHourlyPumpPower), for them we are checking whether or not study is loaded via GUI or via solver in **list.cpp** file: 

![image](https://github.com/AntaresSimulatorTeam/Antares_Simulator/assets/131546833/256e12d4-b4d7-49d0-b9b0-e9a70f88ed69)

guilpier-code replied:
I removed the **bothZero** case because : 
Before calling **EqualizeTSsize** function, we always call a load function (**loadROR**, **loadMINGEN**, ...).
Each of them contain a call to **Matrix::loadFromCSVFile(...)**. And this function always gives to the matrix a width at least equal to one. So, in **EqualizeTSsize(...)**, even the case where one of the 2 matrices has a width = 0 is impossible.
Did I miss something ? Did you have a crash or a bug with GUI while loading hydro TS ?

nikolaredstork replied:
You are absolutely right what you have said, but that is the case when a study is loaded via solver. In case when we are opening (loading) a study via GUI, matrices will not be loaded and we are performing checks for empty matrices. How I know that they are empty when opening study via GUI? I've debug it. This is a snapshot of debug mode, where you can see that ror and storage matrices are empty after loading via GUI. 

![image](https://github.com/AntaresSimulatorTeam/Antares_Simulator/assets/131546833/d36e9ba1-2a69-4793-ad65-d44685d01e59)

I know that we are deprecating this GUI, but should we leave it this way?

guilpier-code replied:
No, you're right, we don't want any bug / crash in GUI, even if it will be removed some day.
Thanks for having tested this.
Correction done.

nikolaredstork replied:
Just to add, this will not crash GUI, but nevertheless it should be considered.

nikolaredstork commented on src/libs/antares/study/area/list.cpp:
```diff
@@ -904,7 +904,11 @@ static bool AreaListLoadFromFolderSingleArea(Study& study,
+            area.hydro.series->EqualizeGenerationTSsizes(area, study.usedByTheSolver, study.gotFatalError);
         }
 
         if (area.hydro.series && study.header.version < 870)
```
Maybe we should extract loading of deprecated files from `AreaListLoadFromFolderSingleArea` function and implement specific load function for deprecated files `AreaLoadFromFolderDeprecatedFiles`, in order to avoid confusion for loading hydro files:

Implementation:
![image](https://github.com/AntaresSimulatorTeam/Antares_Simulator/assets/131546833/277cbf02-5075-4f0d-9c9d-4145bbe331d9)

Function call:
![image](https://github.com/AntaresSimulatorTeam/Antares_Simulator/assets/131546833/d64275b4-d2ea-4c0a-9200-d5acb6cf9fd6)

guilpier-code replied:
The above comment [here](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1707#discussion_r1371459705) should be an answer.

nikolaredstork replied:
Thanks!

nikolaredstork commented on src/libs/antares/study/area/list.cpp:
```diff
@@ -922,14 +926,14 @@ static bool AreaListLoadFromFolderSingleArea(Study& study,
+                area.hydro.series->setHydroModulability(area);
         }
 
+        area.hydro.series->resizeTSinDeratedMode(study.parameters.derated, study.header.version);
```
If we are loading via GUI, matrices ror, storage, mingen, maxHourlyGenPower, maxHourlyPumpPower are not actually loaded and in case of derated mod, loading via GUI code will calculating average column on empty matrix. Resize in derated mod should be called if study is used by the solver.

guilpier-code replied:
Easy to correct. Thanks to have such good eyes.
Correction done.

nikolaredstork commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -224,173 +257,78 @@ void DataSeriesHydro::reset()
 }
 
-bool DataSeriesHydro::postProcessMaxPowerTS(Area& area)
+void DataSeriesHydro::resizeGenerationTS(unsigned int w, unsigned int h)
```
You are you using `resizeGenerationTS` and `resizeRORandSTORAGE` for calculating TS in derated mod? In derated mod, we are calculating average matrices for ror, storage, mingen, maxHourlyGenPower, maxHourlyPumpPower with data provided from .txt files. In this case if derated mod is on matrices will have width 1 filled with junk data (reset function will set all matrix value to 0, resize will only set width and height).

guilpier-code replied:
Oops, it seems I made a mistake. Thanks for seeing that.
Correction done.

nikolaredstork commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -77,128 +133,105 @@ bool DataSeriesHydro::saveToFolder(const AreaName& areaID, const AnyString& fold
+    filepath.clear() << folder << SEP << areaID << SEP << "maxHourlyGenPower.txt";
+    ret = maxHourlyGenPower.loadFromCSVFile(filepath, 1, HOURS_PER_YEAR, &fileContent) && ret;
 
-    timeseriesNumbers.clear();
```
We excluded clearing of timeseriesNumbers matrix?

guilpier-code replied:
you're right. Done.

guilpier-code committed:
Refactor max power TS checks : correct derated mode for hydro

guilpier-code committed:
Refactor max power TS checks : some more corrections due to review

nikolaredstork commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -316,15 +316,24 @@ uint DataSeriesHydro::maxPowerTScount() const
         return;
 
-    resizeGenerationTS(1, HOURS_PER_YEAR);
+    ror.averageTimeseries();
```
I think that ror, storage and mingen should also be averaged if usedBySolver is true and also I think that mingen should be averaged if study header version is greater and equal to 860. Studies below 860 version does not have mingen.txt file, but in constructor of the DataSeriesHydro mingen matrix is set to width 1 with values of 0, meaning that averaging will be performed on the 1x8760 matrix with values of 0.

guilpier-code replied:
ok, done.

nikolaredstork commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -62,6 +62,13 @@ unsigned int EqualizeTSsize(Matrix<double, int32_t>& TScollection1,
+    {
+        TScollection1.reset(1, HOURS_PER_YEAR);
+        TScollection2.reset(1, HOURS_PER_YEAR);
+        return 1;
```
Storage matrix has 365 height (DAYS_PER_YEAR), so we need somehow to parse heights of the matrices.

guilpier-code replied:
ok, done.

guilpier-code committed:
Refactor max power TS checks : some more corrections after review

nikolaredstork commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -153,21 +153,15 @@ bool loadTSfromFile(Matrix<double, int32_t>& ts,
-{
+    bool ret = loadTSfromFile(ror, areaID, folder, "ror.txt", HOURS_PER_YEAR);
+    ret = loadTSfromFile(storage, areaID, folder, "mod.txt", DAYS_PER_YEAR) && ret;
     if (studyVersion < 860)
```
`if(studyVersion >= 860)`
We obviously don't have end-to-end test for minimum generation feature, otherwise that end-to-end test would fail.

guilpier-code replied:
ok, done.

guilpier-code committed:
Refactor max power TS checks : fixing errors introduced by earlier corrections

guilpier-code committed:
Refactor max power TS checks : more corrections after review

guilpier-code committed:
Refactor max power TS checks : more corrections due to tight review :-)

nikolaredstork committed:
Merge branch 'feature/max-hydro-gen-pump-per-h-cr23-rte-comments' into fix/refactor-ts-consistency-ts-after-load

nikolaredstork committed:
Fixing sonar bug

nikolaredstork committed:
Code smells

nikolaredstork committed:
Use init-statement inside the if statement for hydroSeries ptr

nikolaredstork committed:
More code smells
